File: projects/cse_algorithms/problems/HW4/q4_weirdest_programming_language/main.py
MOD = 1000000009

def karatsuba_multiply(a: list[int], b: list[int]) -> list[int]:
    """Fast polynomial multiplication using Karatsuba algorithm."""
    n = len(a)
    if n <= 32:  # Use naive multiplication for small polynomials
        result = [0] * (2 * n - 1)
        for i in range(n):
            for j in range(n):
                result[i + j] = (result[i + j] + a[i] * b[j]) % MOD
        return result
    
    # Split polynomials
    mid = n // 2
    a0, a1 = a[:mid], a[mid:]
    b0, b1 = b[:mid], b[mid:]
    
    # Recursive calls
    z0 = karatsuba_multiply(a0, b0)
    z2 = karatsuba_multiply(a1, b1)
    
    # Compute (a0 + a1)(b0 + b1)
    a_sum = [(a0[i] + a1[i]) % MOD for i in range(len(a0))]
    b_sum = [(b0[i] + b1[i]) % MOD for i in range(len(b0))]
    z1 = karatsuba_multiply(a_sum, b_sum)
    
    # Combine results
    result = [0] * (2 * n - 1)
    for i in range(len(z0)):
        result[i] = z0[i]
    for i in range(len(z2)):
        result[i + 2 * mid] = z2[i]
    for i in range(len(z1)):
        result[i + mid] = (result[i + mid] + z1[i] - z0[i] - z2[i]) % MOD
    
    return result

def solve(g: int, n_total: int, monsters: list[int]) -> int:
    """
    Solve the weirdest programming language problem using optimized polynomial multiplication.
    Theoretical optimal solution: O(g * log n_total)
    
    Args:
        g: Number of gate types
        n_total: Total number of monsters needed
        monsters: List of monsters needed for each gate type
        
    Returns:
        Number of possible sequences modulo 1000000009
    """
    # Sort monsters and remove duplicates
    monsters = sorted(set(monsters))
    
    # Create polynomial coefficients
    coeffs = [0] * (max(monsters) + 1)
    for m in monsters:
        coeffs[m] = 1
    
    # Fast polynomial exponentiation
    def poly_pow(poly: list[int], power: int) -> list[int]:
        if power == 0:
            return [1]
        if power == 1:
            return poly
            
        # Binary lifting with Karatsuba multiplication
        half = poly_pow(poly, power // 2)
        result = karatsuba_multiply(half, half)
        
        if power % 2:
            result = karatsuba_multiply(result, poly)
            
        return result
    
    # Calculate result
    result = poly_pow(coeffs, n_total)
    
    # Return the coefficient of x^n_total
    return result[n_total] if n_total < len(result) else 0

def main():
    import sys
    input_data = sys.stdin.read().split()
    
    # Parse input
    t = int(input_data[0])
    current_index = 1
    
    # Process each test case
    for _ in range(t):
        g = int(input_data[current_index])
        n_total = int(input_data[current_index + 1])
        monsters = [int(x) for x in input_data[current_index + 2:current_index + 2 + g]]
        
        # Solve the problem
        result = solve(g, n_total, monsters)
        
        # Print output
        print(result)
        
        # Update index for next test case
        current_index += 2 + g

if __name__ == "__main__":
    main() 

File: projects/cse_algorithms/problems/HW4/q4_weirdest_programming_language/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MOD 1000000009
#define MAX_MONSTERS 1000
#define MAX_SIZE (1 << 20)  // 2^20 for polynomial size

// Fast modulo multiplication to avoid overflow
static inline long long mul_mod(long long a, long long b) {
    return (a * b) % MOD;
}

// Fast polynomial multiplication using Karatsuba algorithm
void karatsuba_multiply(const long long* a, const long long* b, long long* result, int n) {
    if (n <= 32) {  // Use naive multiplication for small polynomials
        memset(result, 0, (2 * n - 1) * sizeof(long long));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                result[i + j] = (result[i + j] + mul_mod(a[i], b[j])) % MOD;
            }
        }
        return;
    }
    
    int mid = n / 2;
    
    // Allocate temporary arrays on stack for better cache locality
    long long a0[MAX_SIZE/2], a1[MAX_SIZE/2];
    long long b0[MAX_SIZE/2], b1[MAX_SIZE/2];
    long long z0[MAX_SIZE], z1[MAX_SIZE], z2[MAX_SIZE];
    
    // Split polynomials
    memcpy(a0, a, mid * sizeof(long long));
    memcpy(a1, a + mid, (n - mid) * sizeof(long long));
    memcpy(b0, b, mid * sizeof(long long));
    memcpy(b1, b + mid, (n - mid) * sizeof(long long));
    
    // Recursive calls
    karatsuba_multiply(a0, b0, z0, mid);
    karatsuba_multiply(a1, b1, z2, n - mid);
    
    // Compute (a0 + a1)(b0 + b1)
    for (int i = 0; i < mid; i++) {
        a0[i] = (a0[i] + a1[i]) % MOD;
        b0[i] = (b0[i] + b1[i]) % MOD;
    }
    karatsuba_multiply(a0, b0, z1, mid);
    
    // Combine results
    memset(result, 0, (2 * n - 1) * sizeof(long long));
    for (int i = 0; i < 2 * mid - 1; i++) {
        result[i] = z0[i];
    }
    for (int i = 0; i < 2 * (n - mid) - 1; i++) {
        result[i + 2 * mid] = z2[i];
    }
    for (int i = 0; i < 2 * mid - 1; i++) {
        result[i + mid] = (result[i + mid] + z1[i] - z0[i] - z2[i] + 3 * MOD) % MOD;
    }
}

// Fast polynomial exponentiation
void poly_pow(const long long* poly, long long* result, int n, int power) {
    if (power == 0) {
        memset(result, 0, n * sizeof(long long));
        result[0] = 1;
        return;
    }
    if (power == 1) {
        memcpy(result, poly, n * sizeof(long long));
        return;
    }
    
    // Allocate temporary array for intermediate results
    long long temp[MAX_SIZE];
    
    // Binary lifting
    poly_pow(poly, temp, n, power / 2);
    karatsuba_multiply(temp, temp, result, n);
    
    if (power % 2) {
        memcpy(temp, result, n * sizeof(long long));
        karatsuba_multiply(temp, poly, result, n);
    }
}

long long solve(int g, long long n_total, const int* monsters) {
    // Find maximum monster value
    int max_monster = 0;
    for (int i = 0; i < g; i++) {
        if (monsters[i] > max_monster) {
            max_monster = monsters[i];
        }
    }
    
    // Create polynomial coefficients
    long long coeffs[MAX_SIZE] = {0};
    for (int i = 0; i < g; i++) {
        coeffs[monsters[i]] = 1;
    }
    
    // Calculate result
    long long result[MAX_SIZE];
    poly_pow(coeffs, result, max_monster + 1, n_total);
    
    return n_total < MAX_SIZE ? result[n_total] : 0;
}

int main() {
    int t;
    scanf("%d", &t);
    
    while (t--) {
        int g;
        long long n_total;
        scanf("%d %lld", &g, &n_total);
        
        int monsters[MAX_MONSTERS];
        for (int i = 0; i < g; i++) {
            scanf("%d", &monsters[i]);
        }
        
        printf("%lld\n", solve(g, n_total, monsters));
    }
    
    return 0;
} 